### PR TITLE
Remove logging dependency from `SourceKitD`

### PR DESCRIPTION
**Pull request by ahoppen**
Make logging protocol requirements on `SourceKitD` so that implementers of the protocol can decide how to log requests. This allows us to add different kinds of sourcekitd connections that behave differently in the future, for example one that dynamically links against sourcekitd instead of loading it using `dlopen`.

Also rename `SourceKitDImpl` to `DynamicallyLoadedSourceKitD`, for the reasons described above.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

File: Sources/SourceKitD/SourceKitD.swift [hunk 1]
```diff
@@ -83,56 +96,26 @@ extension SourceKitD {
...
-  logger.log(
-    level: (response.error == nil || response.error == .requestCancelled) ? .debug : .error,
-    """
-    Received sourcekitd response:
-    \(response.forLogging)
-    """
-  )
+  public func log(request: SKDRequestDictionary) {}
+  public func log(response: SKDResponse) {}
+  public func log(crashedRequest: SKDRequestDictionary, fileContents: String?) {}
```
Comment (bnbarham) [inline Sources/SourceKitD/SourceKitD.swift:118]: Is it worth having empty defaults?

File: Sources/SourceKitD/DynamicallyLoadedSourceKitD.swift [hunk 2]
```diff
@@ -20,7 +21,7 @@ import struct TSCBasic.AbsolutePath
 ///
 /// Users of this class should not call the api functions `initialize`, `shutdown`, or
 /// `set_notification_handler`, which are global state managed internally by this class.
-public final class SourceKitDImpl: SourceKitD {
+public final class DynamicallyLoadedSourceKitD: SourceKitD {
```
Comment (bnbarham) [inline Sources/SourceKitD/DynamicallyLoadedSourceKitD.swift:24]: `Impl` be gone! Much clearer 👍

File: Sources/SourceKitD/SourceKitD.swift [hunk 3]
```diff
@@ -70,11 +87,7 @@ extension SourceKitD {
   ///   - fileContents: The contents of the file that the request operates on. If sourcekitd crashes, the file contents
   ///     will be logged.
   public func send(_ req: SKDRequestDictionary, fileContents: String?) async throws -> SKDResponseDictionary {
-    logRequest(req)
-
-    let signposter = logger.makeSignposter()
-    let signpostID = signposter.makeSignpostID()
-    let signposterState = signposter.beginInterval("sourcekitd-request", id: signpostID, "Start")
```
Comment (bnbarham) [inline Sources/SourceKitD/SourceKitD.swift:77]: Did you decide the signpost wasn't useful? You could have `logRequest` return some opaque context that `log(response:)` and `log(crashedRequest:)` take if we want to keep it.

**Review by bnbarham — APPROVED** [review]

File: Sources/SourceKitD/SourceKitD.swift (hunk 3 above)
Comment (ahoppen) [reply Sources/SourceKitD/SourceKitD.swift:77]: Yeah. It was a mix of not useful and harder to implement. Just don’t make sense to design something with abstraction layers when it’s doubtful that it provides any real-world value.

File: Sources/SourceKitD/SourceKitD.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SourceKitD/SourceKitD.swift:118]: True, probably better to not have them and force adopters to make decisions about whether they support logging or not.

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Pull request merged**